### PR TITLE
feat(users): excessive data exposure endpoint

### PR DESCRIPTION
**Pull request by Kif-Fando**

### CODE REVIEW

**Comment by derevnjuk** [comment]
@Kif-Fando PR title must satisfy our contributing guide and https://conventionalcommits.org/

**Comment by derevnjuk** [comment]
@Kif-Fando PR title is not just an enumeration of changes, it should reflect the main essence of them (should be limited to 100 symbols)

**Comment by Kif-Fando** [comment]
@derevnjuk please, take a look now

File: src/users/users.service.ts [hunk 1]
```diff
@@ -71,23 +73,31 @@ export class UsersService {
     };
   }
 
-  async findByEmail(email: string): Promise<User> {
-    this.log.debug(`Called findByEmail ${email}`);
-    const user = await this.usersRepository.findOne({ email });
+  async findUser(query: string): Promise<User> {
```
Comment (derevnjuk) [inline src/users/users.service.ts:76]: You should declare a separate method and not mix them (e.g. `findById` and `findByEmail`)

File: src/users/users.service.ts [hunk 2]
```diff
@@ -71,23 +73,31 @@ export class UsersService {
...
+    this.log.debug(`Called findUser ${query}`);
+    const searchData = isNaN(Number(query))
+      ? { email: query }
+      : { id: Number(query) };
+    const user = await this.usersRepository.findOne(searchData);
     if (user) {
       return user;
     } else {
       throw new NotFoundException('User not found');
     }
```
Comment (derevnjuk) [inline src/users/users.service.ts:86]:
```suggestion
    if (!user) {
      throw new NotFoundException('User not found');
    }
    
    return user;
```

File: src/users/users.service.ts [hunk 3]
```diff
@@ -71,23 +73,31 @@ export class UsersService {
...
-    this.log.debug(`Called findById ${id}`);
-    const user = await this.usersRepository.findOne({ id });
-    if (!user) {
-      throw new NotFoundException('User not found');
-    }
-    return user;
+  async searchUsers(query: string): Promise<User[]> {
+    this.log.debug(`Called searchUsers`);
+    const users = await this.usersRepository.findAll();
+    return users.filter((user) => {
```
Comment (derevnjuk) [inline src/users/users.service.ts:92]: Why don't use SQL to select users by WHERE clause?

File: src/users/users.controller.ts [hunk 4]
```diff
@@ -123,10 +125,28 @@ export class UsersController {
...
+
+  @Get('/search/:query')
+  @ApiOperation({
+    description: SWAGGER_DESC_FIND_USERS,
+  })
+  @ApiOkResponse({
+    type: UserDto,
+    description: 'Returns all array with user objects',
+  })
+  async searchUsers(@Param('query') query: string): Promise<FullUserDto[]> {
```
Comment (derevnjuk) [inline src/users/users.controller.ts:145]: Why don't use exact query parameters to filter users via the property name such as `GET /users?firstName=Kiril&company=Bright`

File: src/users/users.controller.ts [hunk 5]
```diff
@@ -123,10 +125,28 @@ export class UsersController {
...
+    }
+  }
+
+  @Get('/search/:query')
+  @ApiOperation({
+    description: SWAGGER_DESC_FIND_USERS,
+  })
+  @ApiOkResponse({
+    type: UserDto,
+    description: 'Returns all array with user objects',
```
Comment (derevnjuk) [inline src/users/users.controller.ts:143]: The wording is unclear
```suggestion
    description: 'Returns the list of users',
```

File: src/users/users.controller.ts [hunk 6]
```diff
@@ -96,25 +98,25 @@ export class UsersController {
...
-      return new UserDto(await this.usersService.findByEmail(email));
+      this.logger.debug(`Find a user by email/id: ${query}`);
+      return new UserDto(await this.usersService.findUser(query));
     } catch (err) {
       throw new HttpException(err.message, err.status);
     }
   }
 
-  @Get('/:id')
+  @Get('/full/:query')
```
Comment (derevnjuk) [inline src/users/users.controller.ts:110]: What does `full` mean?   The meaning and wording are too odd, it is not clear even to me.

File: src/users/users.controller.ts [hunk 7]
```diff
@@ -123,10 +125,28 @@ export class UsersController {
       },
     },
   })
-  async getUserById(@Param('id') id: number): Promise<UserDto> {
+  async getFullUser(@Param('query') query: string): Promise<FullUserDto> {
```
Comment (derevnjuk) [inline src/users/users.controller.ts:128]: ditto

File: src/users/api/FullUserDto.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
+import { ApiHideProperty, ApiProperty } from '@nestjs/swagger';
+import { Exclude } from 'class-transformer';
+
+export class FullUserDto {
```
Comment (derevnjuk) [inline src/users/api/FullUserDto.ts:4]: ditto

Comment (derevnjuk) [inline src/users/api/FullUserDto.ts:4]: `Full UserDto` is a completed duplicate of `UserDto`, there is no difference between them except for some decorators. You have to use `groups` instead

File: src/users/users.controller.swagger.desc.ts [hunk 9]
```diff
@@ -1,8 +1,10 @@
 export const SWAGGER_DESC_OPTIONS_REQUEST = `List of permitted communication options`;
 
-export const SWAGGER_DESC_FIND_USER_BY_EMAIL = `Returns user by email`;
+export const SWAGGER_DESC_FIND_USER = `Returns user by email/id`;
 
-export const SWAGGER_DESC_FIND_USER_BY_ID = `Returns user by id`;
+export const SWAGGER_DESC_FIND_FULL_USER = `Returns full user by email/id`;
+
+export const SWAGGER_DESC_FIND_USERS = `Return users contains text query in firstName/lastName/email/company fields`;
```
Comment (derevnjuk) [inline src/users/users.controller.swagger.desc.ts:7]: The same wording issue

File: src/users/users.controller.ts [hunk 10]
```diff
@@ -78,16 +80,16 @@ export class UsersController {
     this.logger.debug(`Test OPTIONS`);
   }
 
-  @Get('/one/:email')
+  @Get('/one/:query')
```
Comment (derevnjuk) [inline src/users/users.controller.ts:83]: The endpoint can't be variable. It has to accept either ID or email

File: src/users/users.controller.ts [hunk 11]
```diff
@@ -96,25 +98,25 @@ export class UsersController {
       },
     },
   })
-  async getUser(@Param('email') email: string): Promise<UserDto> {
+  async getUser(@Param('query') query: string): Promise<UserDto> {
     try {
-      this.logger.debug(`Find a user by email: ${email}`);
-      return new UserDto(await this.usersService.findByEmail(email));
+      this.logger.debug(`Find a user by email/id: ${query}`);
```
Comment (derevnjuk) [inline src/users/users.controller.ts:103]: ditto

File: src/users/users.controller.swagger.desc.ts [hunk 12]
```diff
@@ -1,8 +1,10 @@
 export const SWAGGER_DESC_OPTIONS_REQUEST = `List of permitted communication options`;
 
-export const SWAGGER_DESC_FIND_USER_BY_EMAIL = `Returns user by email`;
+export const SWAGGER_DESC_FIND_USER = `Returns user by email/id`;
```
Comment (derevnjuk) [inline src/users/users.controller.swagger.desc.ts:3]: ditto

File: src/users/users.controller.swagger.desc.ts [hunk 13]
```diff
@@ -1,8 +1,10 @@
...
 
-export const SWAGGER_DESC_FIND_USER_BY_ID = `Returns user by id`;
+export const SWAGGER_DESC_FIND_FULL_USER = `Returns full user by email/id`;
```
Comment (derevnjuk) [inline src/users/users.controller.swagger.desc.ts:5]: ditto

File: src/users/api/CreateUserRequest.ts [hunk 14]
```diff
@@ -2,6 +2,12 @@ import { ApiProperty } from '@nestjs/swagger';
 import { UserDto } from './UserDto';
 
 export class CreateUserRequest extends UserDto {
+  @ApiProperty()
+  company: string;
+  @ApiProperty()
+  cardNumber: string;
+  @ApiProperty()
+  phoneNumber: string;
```
Comment (derevnjuk) [inline src/users/api/CreateUserRequest.ts:10]: Se comment above

File: public/src/pages/auth/AdminPage.tsx [hunk 15]
```diff
@@ -1,23 +1,52 @@
...
+  };
 
   useEffect(() => {
-    if (user) getAdminStatus(user).then((data) => setIsAdmin(!!data.isAdmin));
-  }, [user]);
+    if (user) {
+      getAdminStatus(user).then((data) => setIsAdmin(!!data.isAdmin));
+      inputValue.length
+        ? searchUsers(inputValue).then((data) => setUsers(data))
+        : setUsers([]);
```
Comment (derevnjuk) [inline public/src/pages/auth/AdminPage.tsx:21]: I'm not sure what you mean here. IMO if the input is empty, you have to display all users

File: public/src/pages/auth/AdminPage.tsx [hunk 16]
```diff
@@ -1,23 +1,52 @@
 import React, { FC, useEffect, useState } from 'react';
-import { getAdminStatus } from '../../api/httpClient';
-import AuthLayout from './AuthLayout';
+import { UserData } from 'src/interfaces/User';
```
Comment (derevnjuk) [inline public/src/pages/auth/AdminPage.tsx:2]: You have to use relative imports

**Review by derevnjuk — CHANGES_REQUESTED** [review]

File: src/users/users.service.ts (hunk 1 above)
Comment (Kif-Fando) [reply src/users/users.service.ts:76]: @derevnjuk AppSecs requested method which searching by email and id the same time

File: src/users/users.controller.ts (hunk 4 above)
Comment (Kif-Fando) [reply src/users/users.controller.ts:145]: Request was on feature which will search through all fields

Comment (derevnjuk) [reply src/users/users.controller.ts:145]: @Kif-Fando I don't see any descriptions of the requested changes. There is no GitHub issue, no PR body. Anyway, it doesn't change the fact that the API design is awful.

File: src/users/users.service.ts (hunk 1 above)
Comment (derevnjuk) [reply src/users/users.service.ts:76]: https://github.com/NeuraLegion/brokencrystals/pull/204#discussion_r960948495

File: public/src/pages/auth/AdminPage.tsx (hunk 15 above)
Comment (Kif-Fando) [reply public/src/pages/auth/AdminPage.tsx:21]: Hm, ok, agreed with you

File: src/users/users.service.ts [hunk 17]
```diff
@@ -71,25 +73,31 @@ export class UsersService {
...
-    } else {
+    if (!user) {
       throw new NotFoundException('User not found');
     }
+    return user;
   }
 
-  async findById(id: number): Promise<User> {
-    this.log.debug(`Called findById ${id}`);
+  async findUserById(id: number): Promise<User> {
```
Comment (derevnjuk) [inline src/users/users.service.ts:85]:
```suggestion
  async findById(id: number): Promise<User> {
```

File: src/users/users.service.ts [hunk 18]
```diff
@@ -71,25 +73,31 @@ export class UsersService {
...
+    this.log.debug(`Called findUserById ${id}`);
     const user = await this.usersRepository.findOne({ id });
     if (!user) {
       throw new NotFoundException('User not found');
     }
     return user;
   }
 
+  async searchUsersByName(query: string): Promise<User[]> {
```
Comment (derevnjuk) [inline src/users/users.service.ts:94]:
```suggestion
  async searchByName(query: string): Promise<User[]> {
```

File: src/users/users.service.ts [hunk 19]
```diff
@@ -71,25 +73,31 @@ export class UsersService {
     };
   }
 
-  async findByEmail(email: string): Promise<User> {
-    this.log.debug(`Called findByEmail ${email}`);
+  async findUserByEmail(email: string): Promise<User> {
```
Comment (derevnjuk) [inline src/users/users.service.ts:76]:
```suggestion
  async findByEmail(email: string): Promise<User> {
```

**Review by derevnjuk — CHANGES_REQUESTED** [review]

**Review by derevnjuk — APPROVED** [review]

**Pull request merged**